sequencer.go

```diff
@@ -7,28 +7,46 @@ package desync
 type SeedSequencer struct {
 	seeds   []Seed
 	index   Index
+	isa     InvalidSeedAction
 	current int
 }
 
 // NewSeedSequencer initializes a new sequencer from a number of seeds.
-func NewSeedSequencer(idx Index, src ...Seed) *SeedSequencer {
+func NewSeedSequencer(idx Index, isa InvalidSeedAction, src ...Seed) *SeedSequencer {
 	return &SeedSequencer{
 		seeds: src,
 		index: idx,
+		isa:   isa,
 	}
 }
 
 // Next returns a sequence of index chunks (from the target index) and the
 // longest matching segment from one of the seeds. If source is nil, no
 // match was found in the seeds and the chunk needs to be retrieved from a
 // store. If done is true, the sequencer is complete.
-func (r *SeedSequencer) Next() (segment IndexSegment, source SeedSegment, done bool) {
+func (r *SeedSequencer) Next() (segment IndexSegment, source SeedSegment, done bool, err error) {
 	var (
 		max     uint64
 		advance = 1
 	)
 	for _, s := range r.seeds {
-		n, m := s.LongestMatchWith(r.index.Chunks[r.current:])
+		// If we already know that this seed is invalid, we just skip it
```

Comment:
You could probably break up this loop into 2 separate loops and get delayed validation, as well as chunk-eviction.
- One loop to collect the longest matches from every seed. Basically the upper part of the existing loop. You'd build a slice of candidates
- Below, one loop to iterate over the candidates, find the longest, perform validation if needed (perhaps you have to add `NeedsValidation()` to the `SeedSegment` interface). If the validation fails you mark the whole seed as invalid and move to the next candidate.

And if you want to go one step further, this split would even allow you to skip bad sections of seeds, while keeping using the seed. Could add that as a 3rd mode of operation, only skip bad candidates, but not invalidate the whole seed. It can be slow, but that's a tradeoff, perhaps network bandwidth is more important and trying to use a damaged seed makes sense in some use-cases.

Reply:
Thanks a lot for the review. I should be able to fix the implementation with your suggestions by tomorrow.

> And if you want to go one step further, this split would even allow you to skip bad sections of seeds, while keeping using the seed. Could add that as a 3rd mode of operation, only skip bad candidates, but not invalidate the whole seed. It can be slow, but that's a tradeoff, perhaps network bandwidth is more important and trying to use a damaged seed makes sense in some use-cases.

This is actually the first thing that I tried but then I decided to implement first the "discard the whole seed" option.
In my first attempt I marked the single chunks that were invalid and kept the rest as is.
While this was working fine for very simple cases, where the invalid file/image had just a few inverted bits, it kinda failed when instead entire chunks where missing or there were additional unexpected chunks.
The remaining chunk hashes were indeed correct, but they were pointing to the wrong blocks, so this required some sort of realigning I guess?

This is a use case that I'm actually quite interested in and wanted to explore more. This is because we have a situation where it's possible to modify a partition that is usually read-only (a "developer" mode that grants you RW permissions).
In this case we may want to push an update on top of an image that has additional files, not available in the index file `.caibx`, or even a few less.
And doing a `desync make` from scratch is quite an expensive operation, that should be avoided if possible.

Reply:
Thanks for making this change. You should be able to extend this quite easily to use as much of the seed as possible and not fail completely. You'd just need a 3rd case checking for something like `InvalidSeedActionContinue` below (with a switch-case).

In another PR of course.